### PR TITLE
ENH: Implement the NumPy C SIMD vectorization interface

### PR DESCRIPTION
**Pull request by seiko2plus**
### This pullrequest changes

---
#### Implement the NumPy C SIMD vectorization interface

"NPYV" or universal intrinsics as NEP-38 define it, are types and functions
 intended to simplify the vectorization of code on different platforms.

The current implementations support  SIMD extensions SSE, AVX2, AVX512,
 VSX and NEON on the top of C definitions that defined by the new
 generated header '_cpu_dispatch.h' which included by 'cpu_dispatch.h'.

And covers the following operations:

 - memory load and store
 - zero, setall, set, select, reinterpret
 - boolean conversions
 - (add, sub, mul, div, adds, subs)
 - logical
 - comparison
 - left and right shifting
 - combine, zip

### CODE REVIEW

File: numpy/core/src/common/simd/neon/misc.h [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+// vector with zero lanes
+#define npyv_zero_u8()  vreinterpretq_u8_s32(npyv_zero_s32())
+#define npyv_zero_s8()  vreinterpretq_s8_s32(npyv_zero_s32())
+#define npyv_zero_u16() vreinterpretq_u16_s32(npyv_zero_s32())
+#define npyv_zero_s16() vreinterpretq_s16_s32(npyv_zero_s32())
+#define npyv_zero_u32() vdupq_n_u32((unsigned)0)
+#define npyv_zero_s32() vdupq_n_s32((int)0)
+#define npyv_zero_u64() vreinterpretq_u64_s32(npyv_zero_s32())
+#define npyv_zero_s64() vreinterpretq_s64_s32(npyv_zero_s32())
+#define npyv_zero_f32() vreinterpretq_f32_s32(npyv_zero_s32())
```
Comment (Qiyu8) [inline numpy/core/src/common/simd/neon/misc.h:17]: can `vdupq_n_f32(0.0f)` replace `vreinterpretq_f32_s32(npyv_zero_s32())`?

Comment (seiko2plus) [reply numpy/core/src/common/simd/neon/misc.h]: sure

**Comment by seiko2plus** [comment]
@mattip, Should I move the SIMD module and testing unit into a separated pr?

**Comment by mattip** [comment]
> Should I move the SIMD module and testing unit into a separated pr?

Yes please.

File: numpy/core/include/numpy/npy_common.h [hunk 2]
```diff
@@ -141,6 +141,14 @@
     #define NPY_INLINE
 #endif
 
+#ifdef _MSC_VER
+    #define NPY_FINLINE static __forceinline
+#elif defined(__GNUC__)
+    #define NPY_FINLINE static NPY_INLINE __attribute__((always_inline))
+#else
+    #define NPY_INLINE static
```
Comment (mattip) [inline numpy/core/include/numpy/npy_common.h:149]:
```suggestion
    #define NPY_FINLINE static
```

**Comment by seiko2plus** [comment]
@mattip, done. I opened a new pr(#16782) contains the SIMD module and testing units

**Review by mattip — COMMENTED** [review]
There is a lot here, very impressive. Is this all new code or was it adapted from somewhere else?

Is there a script we can run on each of the directories to verify that all the universal intrinsics are defined for all the variations?

File: numpy/core/src/common/simd/simd.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+// TODO: Add support for VSX(2.06) and BE Mode
+#if defined(NPY_HAVE_VSX2) && defined(__LITTLE_ENDIAN__)
+    #include "vsx/vsx.h"
+#endif
+
+#ifdef NPY_HAVE_NEON
+    #include "neon/neon.h"
+#endif
+
```
Comment (mattip) [inline numpy/core/src/common/simd/simd.h:46]:
```suggestion
/* Fallback if none of the above have been defined. Disables SIMD features */
```

**Comment by seiko2plus** [comment]
@mattip,
There is a lot here, very impressive. Is this all new code or was it adapted from somewhere else?

No, I wrote it from scratch. I tried to put all my experience on it which mostly I gain it during working on OpenCV.

**Comment by seiko2plus** [comment]
@mattip,
Is there a script we can run on each of the directories to verify that all the universal intrinsics are defined for all the variations?

I'm not sure what you actually mean, but I guess yes, the SIMD module and the testing unit check `test_simd.py` in #16782

File: numpy/core/src/common/simd/simd.h (hunk 3 above)
Comment (seiko2plus) [reply numpy/core/src/common/simd/simd.h:46]: in the extension header, e.g. `avx512/avx512.h`.

**Comment by mattip** [comment]
> the SIMD module and the testing unit check test_simd.py in #16782

Thanks. LGTM.

**Pull request merged**